pkg/cloudhypervisor/client.go

```diff
@@ -0,0 +1,467 @@
+	// AddNetworkDevice will add a new network device to the VM.
+	AddNetworkDevice(ctx context.Context, config *NetConfig) (*PciDeviceInfo, error)
+	// AddVsockDevice will add a new vsock device to the VM.
+	AddVsockDevice(ctx context.Context, config *VsockConfig) (*PciDeviceInfo, error)
+	// AddVdpaDevice will add a new vdpa device to the VM.
+	AddVdpaDevice(ctx context.Context, config *VdpaConfig) (*PciDeviceInfo, error)
+	// Snapshot will create a snapshot of the VM.
+	Snapshot(ctx context.Context, config *VmSnapshotConfig) error
+	// CoreDump will take a core dump of the VM.
+	CoreDump(ctx context.Context, config *VMCoreDumpData) error
+	// Restore will restore a VM from a snapshot.
+	Restore(ctx context.Context, config *RestoreConfig) error
+	// ReceiveMigration will receive a VM migration from a URL.
+	ReceiveMigration(ctx context.Context, config *ReceiveMigrationData) error
+	// SendMigration will send a VM migration to a URL.
+	SendMigration(ctx context.Context, config *SendMigrationData) error
+}
+
+type client struct {
+	builder *requests.Builder
+}
+
+// New will create a new cloud hypervisor client.
+func New(socketPath string) Client {
+	t := &http.Transport{
+		DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+			return net.Dial("unix", socketPath)
+		},
+	}
+
+	return &client{
+		builder: requests.URL(DefaultServerEndpoint).Transport(t),
+	}
+}
+
+// VmmPing checks for API server availability.
+func (c *client) VmmPing(ctx context.Context) (*VmmPingResponse, error) {
+	resp := &VmmPingResponse{}
+
+	if err := c.builder.Clone().Path(PathVmmPing).ToJSON(resp).Fetch(ctx); err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+// VmmShutdown shuts down the cloud-hypervisor vmm.
+func (c *client) VmmShutdown(ctx context.Context) error {
+	return c.builder.Clone().Path(PathVmmShutdown).Put().Fetch(ctx)
+}
+
+// Info returns general information about the cloud-hypervisor Virtual Machine (VM) instance.
+func (c *client) Info(ctx context.Context) (*VmInfo, error) {
+	data := &VmInfo{}
+
+	if err := c.builder.Clone().Path(PathVmInfo).
+		ToJSON(data).
+		Fetch(ctx); err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+// Create will create the cloud-hypervisor Virtual Machine (VM) instance. The instance is not booted, only created
+func (c *client) Create(ctx context.Context, config *VmConfig) error {
+	return c.builder.Clone().Path(PathVmCreate).Put().BodyJSON(config).Fetch(ctx)
+}
+
+// Delete will delete the cloud-hypervisor Virtual Machine (VM) instance.
+func (c *client) Delete(ctx context.Context) error {
+	return c.builder.Clone().Path(PathVmDelete).Put().Fetch(ctx)
+}
+
+// Boot will boot the previously created VM instance.
+func (c *client) Boot(ctx context.Context) error {
+	return c.
+		builder.
+		Clone().
+		Path(PathVmBoot).
+		AddValidator(CustomErrValidator(map[int]string{
+			404: "The VM instance could not boot because it is not created yet",
```

Comment:
would be good to have all these as consts, but happy to leave that as a TODO for later